From 26efa8e08d1e355b0e3bbf18e854919e0ab43fdc Mon Sep 17 00:00:00 2001
From: Michal Klocek <michal.klocek@qt.io>
Date: Thu, 20 Aug 2026 11:36:56 +0200
Subject: [PATCH] Adaptation for CVE-2026-17688
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

According to the fix in Chromium methods OnTouchEvent/OnTouchEventAck
synchronously call back into the client OnGestureEvent which can
potentially end up in  navigation/view teardown. Add 'protectors'
to guard against that.

Pick-to: 6.140 6.11
Task-number: QTBUG-148657
Change-Id: Idd7b73276b0089b078f8bfd54815983285c305cf
Reviewed-by: Michael Brüning <michael.bruning@qt.io>
---
 src/core/render_widget_host_view_qt.cpp | 18 ++++++++++++++----
 src/core/render_widget_host_view_qt.h   |  2 +-
 2 files changed, 15 insertions(+), 5 deletions(-)

diff --git a/src/core/render_widget_host_view_qt.cpp b/src/core/render_widget_host_view_qt.cpp
index cd50ba693d0..ee59f5371b5 100644
--- a/src/core/render_widget_host_view_qt.cpp
+++ b/src/core/render_widget_host_view_qt.cpp
@@ -158,7 +158,8 @@ class FlingingCompositor : public ui::Compositor
 RenderWidgetHostViewQt::RenderWidgetHostViewQt(content::RenderWidgetHost *widget)
     : content::RenderWidgetHostViewBase::RenderWidgetHostViewBase(widget)
     , m_taskRunner(base::SingleThreadTaskRunner::GetCurrentDefault())
-    , m_gestureProvider(QtGestureProviderConfig(), this)
+    , m_gestureProvider(
+              base::MakeRefCounted<ui::FilteredGestureProvider>(QtGestureProviderConfig(), this))
     , m_frameSinkId(host()->GetFrameSinkId())
     , m_delegateClient(new RenderWidgetHostViewQtDelegateClient(this))
 {
@@ -200,6 +201,8 @@ RenderWidgetHostViewQt::RenderWidgetHostViewQt(content::RenderWidgetHost *widget
 
 RenderWidgetHostViewQt::~RenderWidgetHostViewQt()
 {
+    m_gestureProvider->Shutdown();
+
     m_delegate.reset();
 
     QObject::disconnect(m_adapterClientDestroyedConnection);
@@ -859,13 +862,20 @@ void RenderWidgetHostViewQt::ProcessAckedTouchEvent(const input::TouchEventWithL
 {
     const bool eventConsumed = (ack_result == blink::mojom::InputEventResultState::kConsumed);
     const bool isSetBlocking = input::InputEventResultStateIsSetBlocking(ack_result);
-    m_gestureProvider.OnTouchEventAck(touch.event.unique_touch_event_id, eventConsumed, isSetBlocking);
+    // Keep the gesture provider alive during event dispatch as it can trigger
+    // synchronous view destruction.
+    scoped_refptr<ui::FilteredGestureProvider> protector(m_gestureProvider);
+    protector->OnTouchEventAck(touch.event.unique_touch_event_id, eventConsumed, isSetBlocking);
 }
 
 void RenderWidgetHostViewQt::processMotionEvent(const ui::MotionEvent &motionEvent)
 {
-    auto result = m_gestureProvider.OnTouchEvent(motionEvent);
-    if (!result.succeeded)
+    auto weakThis = GetWeakPtr();
+    // Keep the gesture provider alive during event dispatch as it can trigger
+    // synchronous view destruction.
+    scoped_refptr<ui::FilteredGestureProvider> protector(m_gestureProvider);
+    auto result = protector->OnTouchEvent(motionEvent);
+    if (!result.succeeded || !weakThis)
         return;
     blink::WebTouchEvent touchEvent = ui::CreateWebTouchEventFromMotionEvent(motionEvent,
                                                                              result.moved_beyond_slop_region,
diff --git a/src/core/render_widget_host_view_qt.h b/src/core/render_widget_host_view_qt.h
index e935ae888c9..395334f7587 100644
--- a/src/core/render_widget_host_view_qt.h
+++ b/src/core/render_widget_host_view_qt.h
@@ -203,7 +203,7 @@ class RenderWidgetHostViewQt
 
     std::unique_ptr<input::CursorManager> m_cursorManager;
 
-    ui::FilteredGestureProvider m_gestureProvider;
+    scoped_refptr<ui::FilteredGestureProvider> m_gestureProvider;
 
     viz::FrameSinkId m_frameSinkId;
     std::unique_ptr<RenderWidgetHostViewQtDelegateClient> m_delegateClient;
